linkerd/stack/src/queue.rs

```diff
@@ -1,93 +1,153 @@
 use crate::{
     failfast::{self, FailFast},
     layer::{self, Layer},
-    BoxService, CloneParam, ExtractParam, NewService, Service,
+    BoxService, ExtractParam, NewService, Service,
 };
 use linkerd_error::Error;
-use std::{fmt, marker::PhantomData, time::Duration};
+use std::{fmt, marker::PhantomData, time};
 use tower::buffer::Buffer;
 
-#[derive(Debug, Copy, Clone)]
-pub struct QueueConfig {
-    /// The number of requests (or connections, depending on the context) that
-    /// may be buffered
-    pub capacity: usize,
+/// A param type that configures the capacity of a queue.
+#[derive(Clone, Debug)]
+pub struct Capacity(pub usize);
 
-    /// The maximum amount of time a request may be buffered before failfast
-    /// errors are emitted.
-    pub failfast_timeout: Duration,
-}
+/// A param type that configures a queue's failfast timeout.
+#[derive(Clone, Debug)]
+pub struct Timeout(pub time::Duration);
 
-pub struct NewQueue<X, Req, N> {
+/// A `NewService` that builds buffered inner services.
+pub struct NewQueue<Z, X, Req, N> {
     inner: N,
     extract: X,
-    _req: PhantomData<fn(Req)>,
+    _req: PhantomData<fn(Req, Z)>,
 }
 
-pub type Queue<Req, Rsp> = failfast::Gate<Buffer<BoxService<Req, Rsp, Error>, Req>>;
+/// A marker type to indicate a queue applies failfast timeouts.
+#[derive(Debug)]
+pub enum WithTimeout {}
+
+pub type NewQueueWithTimeout<X, Req, N> = NewQueue<WithTimeout, X, Req, N>;
+
+pub type QueueWithTimeout<Req, Rsp, E = Error> =
+    failfast::Gate<Buffer<BoxService<Req, Rsp, E>, Req>>;
+
+/// A marker type to indicate a queue does not apply a timeout.
+#[derive(Debug)]
+pub enum WithoutTimeout {}
+
+pub type NewQueueWithoutTimeout<X, Req, N> = NewQueue<WithoutTimeout, X, Req, N>;
+
+pub type QueueWithoutTimeout<Req, Rsp, E = Error> = Buffer<BoxService<Req, Rsp, E>, Req>;
```

Comment:
what's the advantage of using the marker types versus just having separate `NewQueue`/`NewQueueWithTimeout` structs?

if `NewQueueWithTimeout` and `NewQueueWithoutTimeout` were just separate structs rather than using the marker, we could invoke the constructors as `NewQueueWithTimeout::layer`/`NewQueue::layer`...

Reply:
yeah agreed. I've left `NewQueue` to _include_ the timeout, since that's generally the behavior we want. We should be explicit when that safety measure isn't in place.